extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -59,34 +70,73 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
                     public CompletionStage<Boolean> apply(Permission permission) {
                         return securityIdentity.checkPermission(permission);
                     }
-                })
-                .build();
+                }).build();
     }
 
     public CompletionStage<SecurityIdentity> authenticate(RoutingContext context,
             IdentityProviderManager identityProviderManager,
             DefaultTenantConfigResolver resolver) {
         Cookie sessionCookie = context.request().getCookie(SESSION_COOKIE_NAME);
+        TenantConfigContext configContext = resolver.resolve(context, true);
 
         // if session already established, try to re-authenticate
         if (sessionCookie != null) {
             String[] tokens = sessionCookie.getValue().split(COOKIE_DELIM);
-            return authenticate(identityProviderManager, new IdTokenCredential(tokens[0], context))
-                    .thenCompose(new Function<SecurityIdentity, CompletionStage<SecurityIdentity>>() {
+            String idToken = tokens[0];
+            String accessToken = tokens[1];
+            String refreshToken = tokens[2];
+
+            return authenticate(identityProviderManager, new IdTokenCredential(idToken, context))
+                    .handle(new BiFunction<SecurityIdentity, Throwable, SecurityIdentity>() {
                         @Override
-                        public CompletionStage<SecurityIdentity> apply(SecurityIdentity securityIdentity) {
-                            return CompletableFuture
-                                    .completedFuture(augmentIdentity(securityIdentity, tokens[1], tokens[2], context));
+                        public SecurityIdentity apply(SecurityIdentity identity, Throwable throwable) {
+                            if (isLogout(context, configContext)) {
+                                // force logout, the challenge will be built accordingly and redirect to the logout endpoint at the OP
+                                throw new AuthenticationFailedException();
+                            }
+
+                            // if authentication failed, the reason could be the token no longer being valid, so we try a 
+                            // silent refresh if required
+                            if (throwable != null) {
```

Comment:
@pedroigor should we have a check here that the token has expired ? At the moment the way to do is to check the exception message, Vertx reports `JWT is expired` or something like that. I'm going to work on a few Vertx PRs soon enough and perhaps we can introduce a dedicated exception too.

Reply:
There is an assumption that the attempt to refresh should only be done if there is an existing session (e.g.: cookie exists). I'm missing this check too.

That said, exceptions thrown should be only related to expiration, AFAIK. That is why I'm not checking for a specific exception.

I'm OK about being more defensive and do something along the line you suggested, but if you think my assumption is correct, maybe unnecessary to add a new exception ?

Reply:
@pedroigor The verification can definitely fail if the token has an invalid signature, right ?

Reply:
It would be hard to have the signature invalid considering that the session cookie was previously established. Like I said, it is OK to have a specific exception for this, but I tried to keep it simple and not introduce more exceptions considering that we should get expiration errors 99% of time, I think.

Reply:
@pedroigor But we can't afford trying to refresh a token which someone replaced in a cookie ? It is not hard to do at all for anyone spending their time hacking the cookies. We don't have a dedicated exception yet anyway, but as I said we can check the text of the message

Reply:
Well, if you really want to be more explicit about the exception. Then I would say we should add proper exceptions.

Reply:
The dedicated exceptions can make things cleaner but we already know Vertx will throw the exception with a specific message indicating the expiry, IMHO we can't start refreshing unless it is an expiry related cause